pyoptsparse/meson.build

```diff
@@ -1,69 +1,56 @@
-# NumPy include directory - needed in all submodules
-incdir_numpy = get_option('incdir_numpy')
-if incdir_numpy == ''
-    incdir_numpy = run_command(py3_target,
-      [
-        '-c',
-        'import os; os.chdir(".."); import numpy; print(numpy.get_include())'
-      ],
-      check: true
-    ).stdout().strip()
+# <!-- from https://github.com/scipy/scipy/blob/4d9f5e65af06d4cc3f770407f1a66a185675eea9/scipy/meson.build
+
+incdir_numpy = meson.get_external_property('numpy-include-dir', 'not-given')
+if incdir_numpy == 'not-given'
+  incdir_numpy = run_command(py3,
+    [
+      '-c',
+      '''import os
+import numpy as np
+try:
+  incdir = os.path.relpath(np.get_include())
+except Exception:
+  incdir = np.get_include()
+print(incdir)
+  '''
+    ],
+    check: true
+  ).stdout().strip()
+
+  # We do need an absolute path to feed to `cc.find_library` below
+  _incdir_numpy_abs = run_command(py3,
+    ['-c', 'import os; os.chdir(".."); import numpy; print(numpy.get_include())'],
+    check: true
+  ).stdout().strip()
+else
+  _incdir_numpy_abs = incdir_numpy
 endif
-# this creates a raw string which is useful for Windows use of '\' for paths
-incdir_numpy = '''@0@'''.format(incdir_numpy)
-
-# HACK: Meson prefixes filenames of intermediate compiled objects with their filepath. This poses a problem for conda builds
-# since conda ensures the host environment directory has 255 characters so the meson object filenames then exceed 255
-# characters. To remedy this, the fortranobject.c file from numpy is copied into pyoptsparse so that the meson build
-# uses a relative path, rather than an absolute path, thus reducing the auto generated object filename
-# see for example https://github.com/mesonbuild/meson/issues/4226
-run_command(py3_command,
-  [
-    '-c',
-    'import os; os.chdir(".."); import shutil; shutil.copy(os.path.join(r"' + incdir_numpy + '", "..", "..", "f2py", "src", "fortranobject.c"), "pyoptsparse")'
-  ],
-  check: true
-)
-
 inc_np = include_directories(incdir_numpy)
 
+# Don't use the deprecated NumPy C API. Define this to a fixed version instead of
+# NPY_API_VERSION in order not to break compilation for released SciPy versions
+# when NumPy introduces a new deprecation.
+numpy_nodepr_api = ['-DNPY_NO_DEPRECATED_API=NPY_1_9_API_VERSION']
+np_dep = declare_dependency(include_directories: inc_np, compile_args: numpy_nodepr_api)
 
-# TODO: pyoptsparse supports numpy>=1.16 but numpy.f2py.get_include() wasnt added until later, raise numpy version?
-#incdir_f2py = run_command(py3_target,
-#  [
-#    '-c',
-#    'import os; os.chdir(".."); import numpy.f2py; print(numpy.f2py.get_include())'
-#  ],
-#  check : true
-#).stdout().strip()
 incdir_f2py = incdir_numpy / '..' / '..' / 'f2py' / 'src'
```

Comment:
We can probably call `numpy.f2py.get_include()` here in case they ever change path